Pass site url as callback parameter to enable fetching Push token against external tenants

AndreiZe opened the pull request:
In order for user to load Fluid content from a file which is stored in external O365 tenant it is necessary to request OAuth token with external tenant explicitly specified as part of AAD authority that should issue a token. 

StorageTokenFetcher callback signature includes siteUrl parameter which can be used to determine AAD tenant. We need similar capability for PushTokenFetcher callback. This is required because Push token must be able to establish collab session within the context of external tenant. 

Note that this is breaking change, this is intentional to ensure that token fetcher implementations think about external access case and either provide proper support for it or return error in case external access is not supported.

I manually validated the part pertaining to handling auth failure where access token was not issues by expected authority. In this case www-authenticate header is returned with 'realm' property indicating AAD tenant authority that should be used for issuing access token.

Commit:
Pass site url as callback parameter to enable fetching Push token against external tenants

Commit:
Fixing logic for detecting AAD tenant hostname based on sharepoint tenant hostname

sumedhb1995 commented on packages/utils/odsp-doclib-utils/src/odspDocLibUtils.ts:
```diff
@@ -13,11 +13,17 @@ export function isSharepointURL(server: string) {
     }
-    return tenant === "microsoft" ? "organizations" : `${tenant}.onmicrosoft.com`;
+
+    let restOfTenatHostname = server.substr(tenantName.length).toLowerCase();
```
nit: spelling of tenant
```suggestion
    let restOfTenantHostname = server.substr(tenantName.length).toLowerCase();
```

AndreiZe replied:
Thanks for catching, I fixed it as part of another refactoring that I published in 3rd iteration of this PR

sumedhb1995 commented on packages/utils/tool-utils/src/odspTokenManager.ts:
```diff
@@ -46,9 +46,7 @@ export type OdspTokenConfig = {
-export interface IPushCacheKey { isPush: true }
-export interface IOdspCacheKey { isPush: false; server: string }
-export type OdspTokenManagerCacheKey = IPushCacheKey | IOdspCacheKey;
+export interface OdspTokenManagerCacheKey { isPush: boolean; server: string; }
```
nit: since this is now an interface, we might want to name it starting with I
```suggestion
export interface IOdspTokenManagerCacheKey { isPush: boolean; server: string; }
```

AndreiZe replied:
I fixed it as part of another refactoring that I published in 3rd iteration of this PR

vladsud commented on BREAKING.md:
```diff
@@ -1,3 +1,10 @@
+- [Extend PushTokenFetcher callback signature](#extend-pushtokenfetcher-callback-signature-to-include-siteurl-parameter)
+
+## Extend PushTokenFetcher callback signature to include siteUrl parameter
+This will make it possible to acquire Push token that can be used with Fluid ODSP documents stored externally.
```
We either need to replace PUSH with "ordering service" or add "ODSP driver's ..." to it, as "PUSH" has no clear meaning for general audience

AndreiZe replied:
@vladsud, is it okay to keep PushTokenFetcher as type name or should it be renamed as well? Note that this type is defined within odsp-driver package so I am thinking it might be okay  to keep it unchanged but I am open to suggestions.

vladsud replied:
I think it's fine to have it as is. It's just the word "PUSH" ideally should not leave ODSP driver packages, as people reading it (including breaking.md) will not be able to connect the dots...

AndreiZe replied:
I went ahead with deeper refactoring and changed description appropriately.

vladsud commented on packages/drivers/odsp-driver/src/odspDocumentServiceFactoryCore.ts:
```diff
@@ -162,13 +162,15 @@ export class OdspDocumentServiceFactoryCore implements IDocumentServiceFactory {
                 logger,
                 { eventName: "GetWebsocketToken" },
-                async (event) => tokenFetcher(options.refresh, options.claims).then((tokenResponse) => {
+                async (event) => tokenFetcher(resolvedUrl.siteUrl, options.refresh, options.claims)
```
Maybe time to create an interface and pass as single object? :)

AndreiZe replied:
Agreed. I went ahead and made a few more changes to streamline type definitions that hopefully make things easier to understand and consume.

Commit:
Refactor token fetch types

Commit:
Fixed lint error

vladsud commented:
@AndreiZe , do you need help merging it? Do you have permissions?

Commit:
Minor comments

Commit:
Parse auth error to extract tenant id that should be used when fetching access token

Commit:
REname types and interface

AndreiZe commented:
> @AndreiZe , do you need help merging it? Do you have permissions?

@vladsud , I needed to take a pause as I wanted to validate a few assumptions before committing to specific approach. I have added a few more changes, most notably, I added support for retrying auth errors which report back the tenant that should b used to issue tokens. I also did some type consolidation. Please take another look.

And, yes, I do need help with merging this as I don't have rights to do so on my own.

Commit:
Fixed merge conflicts

markfields commented on packages/utils/odsp-doclib-utils/src/odspAuth.ts:
```diff
@@ -50,14 +50,12 @@ export function getFetchTokenUrl(server: string): string {
 }
 
 export function getLoginPageUrl(
-    isPush: boolean,
```
@AndreiZe this is the function from the build error - you removed one of the parameters 😋

AndreiZe replied:
Thanks for pointing this out! I've been looking at wrong place before

Commit:
Fix build break

markfields commented on packages/drivers/odsp-driver/src/tokenFetch.ts:
```diff
@@ -24,54 +24,70 @@ export interface TokenFetchOptions {
+    type: "OneDrive";
+}
+
+export type SharingLinkTokenFetchOptions = GraphResourceTokenFetchOptions | OneDriveResourceTokenFetchOptions;
```
Are the Graph and OneDrive types used directly? If not maybe just have this extend `OdspResourceTokenFetchOptions` directly with `type: "Graph" | "OneDrive"`

AndreiZe replied:
I think of reworking this whole share link functionality to simplify things. One thing that is stopping me is that there is a talk of moving share link generation to server. In any case I expect some churn here in the future so will keep things as is for now.

markfields commented on packages/drivers/odsp-driver/src/tokenFetch.ts:
```diff
@@ -24,54 +24,70 @@ export interface TokenFetchOptions {
  * whether token came from cache. Legacy implementation may return a string for token value;
  * in this case it should be assumes that fromCache signal is undefined. Null is returned in case of failure.
  */
-export type SharingLinkTokenFetcher =
-    (siteUrl: string, scopeFor: SharingLinkScopeFor, refresh: boolean, claims?: string)
-        => Promise<string | TokenResponse | null>;
```
Why is this going away?

markfields commented on packages/drivers/odsp-driver/src/tokenFetch.ts:
```diff
@@ -24,54 +24,70 @@ export interface TokenFetchOptions {
  * in this case it should be assumes that fromCache signal is undefined. Null is returned in case of failure.
  */
-export type PushTokenFetcher = (refresh: boolean, claims?: string) => Promise<string | TokenResponse | null>;
+export type OdspResourceTokenFetcher =
```
Idea: define a single `TokenFetcher<T>` type where `T` extends `TokenFetchOptions` instead of indivdual `FooFetcher` types?

AndreiZe replied:
Good idea

markfields commented on packages/drivers/odsp-driver/src/test/odspError.spec.ts:
```diff
@@ -203,20 +203,68 @@ describe("Odsp Error", () => {
         const res = await getWithRetryForTokenRefresh(async (options) => {
             if (
                 options.refresh &&
                 options.claims === "{\"access_token\":{\"nbf\":{\"essential\":true, \"value\":\"1597959090\"}}}"
```
Do you also want to make sure the tenantId was plumbed through?

markfields replied:
Or assert it's NOT there if that's expected...? I see the tests below now  Maybe that's silly, your call.

markfields commented on BREAKING.md:
```diff
@@ -1,10 +1,15 @@
+- [Removed some api implemenations from odsp-driver](#Removed-some-api-implemenations-from-odsp-driver)
+- [Refactored token fetcher types in odsp-driver](#refactored-token-fetcher-types-in-odsp-driver)
 
 ### Removed-some-api-implemenations-from-odsp-driver
```
While you're here - `implemenation` is misspelled

AndreiZe replied:
done

markfields reviewed (commented):
This is a nice change! My one concern is test coverage for all the refactored parts.  The more UTs you can add to cover the refactored bits the better, and if that's not possible/feasible at the moment, please take care to manually walk through these various codepaths, or confirm they have coverage elsewhere (e.g. the `nodeStressTest` - and very soon all e2e tests will run against ODSP too!)

Commit:
Additional changes for getSharepointTenant logic

Commit:
Fix build break due to merge conflict

AndreiZe commented on packages/drivers/odsp-driver/src/graph.ts:
```diff
@@ -114,8 +121,7 @@ export async function getShareLink(
 export async function getShareLinkCore(
-    getShareLinkToken:
-        (options: TokenFetchOptions, scopeFor: SharingLinkScopeFor, siteUrl: string) => Promise<string | null>,
+    getToken: (options: SharingLinkTokenFetchOptions) => Promise<string | TokenResponse | null>,
```
No, it has to be SharingLinkTokenFetchOptions

AndreiZe replied:
I will make it SharingLinkTokenFetcher

AndreiZe commented on packages/drivers/odsp-driver/src/odspDocumentServiceFactoryWithCodeSplit.ts:
```diff
@@ -7,14 +7,14 @@ import { IDocumentServiceFactory } from "@fluidframework/driver-definitions";
     constructor(
-        getStorageToken: StorageTokenFetcher,
-        getWebsocketToken: PushTokenFetcher,
+        getStorageToken: OdspResourceTokenFetcher,
+        getWebsocketToken: OdspResourceTokenFetcher,
```
Probably keeping them separate is okay. I am not sure if signature would ever deviate in future. This conceptually looks more and more like generic TokenProvider signature but I think it would be better taken as separate change.

AndreiZe commented on packages/drivers/odsp-driver/src/odspError.ts:
```diff
@@ -29,13 +30,19 @@ export function throwOdspNetworkError(
         response ? numberFromHeader(response.headers.get("retry-after")) : undefined, // seconds
-        claims);
+        claims,
+        tenantId);
```
tenant id is OII and it is okay to log.

AndreiZe commented on packages/drivers/odsp-driver/src/vroom.ts:
```diff
@@ -19,7 +19,7 @@ import { EpochTracker } from "./epochTracker";
  * @param method - The type of request, such as GET or POST
  * @param logger - A logger to use for this request
- * @param getVroomToken - A function that is able to provide the vroom token for this request
+ * @param getStorageToken - A function that is able to provide the access token for this request
```
Will do

AndreiZe commented on packages/utils/odsp-doclib-utils/src/odspDocLibUtils.ts:
```diff
@@ -13,11 +13,17 @@ export function isSharepointURL(server: string) {
 }
 
 export function getSharepointTenant(server: string) {
```
Added unit tests. Note that this function will not work for all types of ODSP tenant urls. If this is meant to be sample code then it is okay. if it is meant to work in all situations then there is much more work to be done.

Commit:
Minor additions to address PR feedback

Commit:
Adding missing import

Commit:
Reverse submodule reference change

Commit:
Renaming util methods to convey appropriate semantics

Commit:
Fix unit tests

AndreiZe commented:
@vladsud , @markfields , @jatgarg , @sumedhb1995 , I addressed all comments. Can someone take another look and pull in?

markfields commented:
@AndreiZe can you put some notes in here about what manual testing you did?  This is quite a spanning refactor.  Just to have for the record.  Other than that I think it's ready to merge, will follow up with a few others to confirm.

markfields reviewed (approved):
🎂 LGTM
